onnxruntime/core/optimizer/qdq_transformer/qdq_propagation.cc

```diff
@@ -1,139 +1,277 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-#include "qdq_propagation.h"
+#include "core/optimizer/qdq_transformer/qdq_propagation.h"
 
-#include <deque>
+#include <optional>
 
 #include "core/graph/graph_utils.h"
 #include "core/optimizer/initializer.h"
 #include "core/optimizer/qdq_transformer/qdq_util.h"
 #include "core/optimizer/utils.h"
 
 namespace onnxruntime {
+namespace {
+// like graph_utils::GraphEdge, but also allowing the source end to be an input and the destination end to be an output
+struct ExtendedGraphEdge {
+  enum class End { Source,
```

Comment:
Should this go in graph_utils? Seems fairly general purpose and other optimizers may be able to benefit from it in the future.

Reply:
moved it to a separate file and put it in the graph_utils namespace